lib/pushbox.js

```diff
@@ -0,0 +1,122 @@
+    last: isA.boolean().required(),
+    index: isA.number().required(),
+    messages: isA.array().items(isA.object({
+      index: isA.number().required(),
+      data: isA.string().regex(validators.URL_SAFE_BASE_64).required(),
+    })).optional(),
+  }),
+  // limit == 0 only returns the last index.
+  isA.object({
+    index: isA.number().required(),
+  }),
+  // error case
+  isA.object({
+    error: isA.string().required(),
+    status: isA.number().required(),
+  })
+)
+
+module.exports = function (log, config) {
+  const pool = new Pool(config.pushboxUrl, { timeout: 15000 })
+  const SafeUrl = require('./safe-url')(log)
+
+  return {
+    /**
+     * Retrieves messages for a specific device. Relays the request to Pushbox.
+     *
+     * @param {String} uid - Firefox Account uid
+     * @param {String} deviceId
+     * @param {Object} options
+     * @param {String} oAuthToken - Hex string of the user's Pushbox bearer token
+     * @param {Number} limit
+     * @param {String} [index]
+     * @returns {Promise}
+     */
+    retrieve (uid, deviceId, oAuthToken, limit, index) {
+      log.trace({
+        op: LOG_OP_RETRIEVE,
+        uid,
+        deviceId,
+        index,
+        limit
+      })
+      const query = {
+        limit: limit.toString()
+      }
+      if (index) {
+        query.index = index
+      }
+      const path = new SafeUrl('/v1/store/:service/:uid/:deviceId')
+      return pool.get(path, {service: SERVICE_FXA, uid, deviceId},
+                      {query, headers: { Authorization: `Bearer ${oAuthToken}` }})
+      .then(body => {
+        const d = P.defer()
+        PUSHBOX_RETRIEVE_SCHEMA.validate(body, (err, body) => {
+          if (err) {
+            return d.reject(new Error('Invalid Pushbox response'))
+          }
+          return d.resolve(body)
+        })
+        return d.promise
+      })
+    },
+
+    /**
+     * Store a message for a specific device
+     *
+     * @param {String} uid - Firefox Account uid
+     * @param {String} deviceId
+     * @param {string} topic
+     * @param {String} data - Base64 string of the payload
+     * @param {String} oAuthToken - Hex string of the user's Pushbox bearer token
+     * @returns {Promise} direct url to the stored message
+     */
+    store (uid, deviceId, topic, data, oAuthToken) {
+      const ttl = TOPICS_TTL[topic]
+      log.trace({
+        op: LOG_OP_STORE,
+        uid,
+        deviceId,
+        topic
+      })
+      const path = new SafeUrl('/v1/store/:service/:uid/:deviceId')
+      const params = {service: SERVICE_FXA, uid, deviceId}
+      return pool.post(path, {service: SERVICE_FXA, uid, deviceId}, {
+        data,
+        ttl
+      }, {headers: { Authorization: `Bearer ${oAuthToken}` }})
+      .then(({index}) => {
+        const msgPath = path.render(params)
+        const url = new URL(msgPath, config.pushboxUrl)
+        url.searchParams.set('index', index)
+        url.searchParams.set('limit', 1)
+        return url.href
```

Comment:
Aha, actually I see this is a URL that points to fxa-auth-server, not a direct URL to the message in pushbox.